## Python Tutorial: Basics of Computer Memory and Discrete Math for Beginner Programming

In [1]:
pip install nbconvert[webpdf]

Note: you may need to restart the kernel to use updated packages.


## 1. Computer Memory

Computer memory is an essential component of any computer system. It stores data and instructions used by the CPU (Central Processing Unit) to perform tasks. There are two main types of memory:

RAM (Random Access Memory): This is temporary memory that stores data while the computer is running. RAM is volatile, meaning that it loses its content when the computer is powered off.

ROM (Read Only Memory): This is permanent memory that contains data and instructions required for the computer to boot up and function. Unlike RAM, ROM is non-volatile and retains its content even when the computer is powered off.

Creating a Music-themed Variable (run the below code by clicking or arrow-ing to the cell, and hit either (shift + enter) or (ctrl + enter))

In [1]:
# This is a comment. It's any part preceded by a #, and the computer won't run it.
# It's mostly to help programmers write notes and document code chunks
song_duration = 180  # Duration in seconds
song_duration

180

Above, we output the value of the variable by just writing the name of it below the assignment, and executing the cell; however, this only works in a notebook like this. If you want to output that "song_duration" variable in a regular python program, you do it like so:

In [2]:
print(song_duration)

180


This is called a function call. Basically, there are a lot of pre-made tools in coding that do most of the hard stuff for us. If we wrote this in something like C++ instead of modern Python, it would look like this mess (won't execute, because this notebook is setup for python, not C++):

In [3]:
std::cout << song_duration << std::endl;

SyntaxError: invalid syntax (1024941744.py, line 1)

Now, since it's a variable, which is a place in the computer's RAM cut out for this program, we can alter the variable, and print the same "print" call and it'll reflect the change. Think of the variable as a box in a warehouse, and the contents are the number or value we store in it, but we also carve a name or designator on that box so we can keep track of it and find it when we need it's contents. But we find it by that name on the outside, not the contents inside, because just like a box in a warehouse, we can't see what's inside without finding the box first.

In [4]:
song_duration += 30
print(song_duration)

210


Now, that '+=' operator probably looks weird, but it's just a shortcut for the expression below:

In [5]:
song_duration = song_duration + 30
print(song_duration)

240


You may have noticed just now that the variable "song_duration" didn't output the same value as it just did above (if you're executing these in order without back-tracking and re-executing cells, that is). 

That's because, again, that variable is a place in memory. The first += updated 180 to 210 by adding 30, and the song_duration + 30 updated that same place in memory by adding 30, only the place in memory was now holding 210 instead of 180 because we already did work on that variable (i.e. place in memory).

In programming, we do have a lot of cases where the memory update doesn't stick unless we tell it to though. Like, if we run this line, and then print the variable song_duration again, nothing will happen, because we had the computer perform the operation on that place in memory, but we didn't tell it to update it by saving it with a reassignment (either '=', '+=', or '-=').

In [6]:
song_duration + 30
print(song_duration)

240


## 2. Discrete Math

Discrete math deals with countable, distinct sets of objects or values. It is essential in computer science, as computers work with discrete values (like integers or binary digits) rather than continuous values.

## Sets
A set is a collection of distinct elements. In Python, you can represent sets using the set data structure. Let's create a set of musical notes:

In [7]:
musical_notes = {"C", "D", "E", "F", "G", "A", "B"}
print(musical_notes)

{'F', 'C', 'G', 'E', 'D', 'B', 'A'}


## Functions
A function is a relation that maps input values to output values. In Python, you can define functions using the def keyword. Let's create a simplified function that calculates the frequency of a musical note:


In [8]:
def frequency_of_note(note):
    if note == "A":
        return 440
    elif note == "B":
        return 494
    elif note == "C":
        return 523
    elif note == "D":
        return 587
    elif note == "E":
        return 659
    elif note == "F":
        return 698
    elif note == "G":
        return 784
    else:
        return None

note = "C"
note_frequency = frequency_of_note(note)
print(f"The frequency of {note} is {note_frequency} Hz")

The frequency of C is 523 Hz


The function frequency_of_note takes a musical note as input and returns its approximate frequency in Hz. You can now use this function to find the frequency of any musical note in the set musical_notes.

(Note, that in this format for what we call a "string" of text, they start it with 'f'. This is only needed if you're using a variable value in the flow of that string. So print("Hello World"), would not require an 'f' prefix. But using this allows us to insert variables within {} in the flow of the string output. This is useful when looping through a lot of lines of data and wanting the loop to output it in a certain statement, like:

The frequency of C is 523 Hz
The frequency of G is 784 Hz
The frequency of D is 587 Hz
The frequency of A is 440 Hz
etc.

In [9]:
circle_of_fifths = ['C', 'G', 'D', 'A', 'E', 'B', 'F']
for note in circle_of_fifths:
    note_frequency = frequency_of_note(note)
    print(f"The frequency of {note} is {note_frequency} Hz")

The frequency of C is 523 Hz
The frequency of G is 784 Hz
The frequency of D is 587 Hz
The frequency of A is 440 Hz
The frequency of E is 659 Hz
The frequency of B is 494 Hz
The frequency of F is 698 Hz


## 3. Data Structures

ata structures are ways to organize and store data in a computer's memory. Python provides built-in data structures like lists, dictionaries, and sets. Let's explore them using a music theme.

## Lists
A list is an ordered collection of elements. You can store multiple items in a list and access them using their index. Let's create a list of popular genres:

In [10]:
music_genres = ["Rock", "Pop", "Hip-Hop", "Jazz", "Classical"]
print(music_genres)

['Rock', 'Pop', 'Hip-Hop', 'Jazz', 'Classical']


To access an element in the list, use its index:

In [11]:
first_genre = music_genres[0]
print(f"The first genre in the list is {first_genre}")

The first genre in the list is Rock


The index was something I covered in the 'software\ lessons.txt' file I sent.

## Dictionaries
A dictionary is a collection of key-value pairs. You can store data as key-value pairs and access the values using their keys. Let's create a dictionary of instruments and their families:

In [12]:
instruments = {"Guitar": "Strings", "Drums": "Percussion", "Trumpet": "Brass", "Flute": "Woodwind"}
print(instruments)

{'Guitar': 'Strings', 'Drums': 'Percussion', 'Trumpet': 'Brass', 'Flute': 'Woodwind'}


To access the value associated with a key, use the key:

In [13]:
guitar_family = instruments["Guitar"]
print(f"The guitar belongs to the {guitar_family} family")

The guitar belongs to the Strings family


## 4. Conditional Statements

Conditional statements allow you to execute code based on certain conditions. The if, elif, and else keywords are used to create conditional statements in Python. Let's create a simple function that tells you if a song is short, average, or long based on its duration:

In [14]:
def song_length(duration):
    if duration < 180:
        return "short"
    elif duration >= 180 and duration < 300:
        return "average"
    else:
        return "long"

song_duration = 200
song_length_description = song_length(song_duration)
print(f"The song with a duration of {song_duration} seconds is {song_length_description}.")

The song with a duration of 200 seconds is average.


## 5. Loops
Loops are used to repeat a block of code multiple times. In Python, you can use for loops to iterate over a sequence, like a list or a range of numbers. Let's iterate over the music_genres list and print each genre:

In [15]:
for genre in music_genres:
    print(genre)

Rock
Pop
Hip-Hop
Jazz
Classical


You can also use while loops to repeat code as long as a condition is true. Let's create a loop that prints the first five numbers of the Fibonacci sequence:

In [16]:
a, b = 0, 1
count = 0

print("First 5 numbers of the Fibonacci sequence:")
while count < 5:
    print(a)
    a, b = b, a + b
    count += 1


First 5 numbers of the Fibonacci sequence:
0
1
1
2
3


These concepts form the foundation of programming in Python. Practice using variables, data structures, functions, conditional statements, and loops to become more comfortable with the language. Happy coding!